src/VisualStudio/Core/Def/LanguageService/AbstractPackage.cs

```diff
@@ -27,16 +27,20 @@ internal IComponentModel ComponentModel
 
     protected override async Task InitializeAsync(CancellationToken cancellationToken, IProgress<ServiceProgressData> progress)
     {
-        await base.InitializeAsync(cancellationToken, progress).ConfigureAwait(true);
+        // Should only be called from a threadpool. Opinionated, as package load sequence thread switches are impactful.
+        Contract.ThrowIfTrue(JoinableTaskFactory.Context.IsOnMainThread);
 
-        await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+        await base.InitializeAsync(cancellationToken, progress).ConfigureAwait(true);
 
-        _componentModel_doNotAccessDirectly = (IComponentModel?)await GetServiceAsync(typeof(SComponentModel)).ConfigureAwait(true);
+        _componentModel_doNotAccessDirectly = await GetServiceAsync<SComponentModel, IComponentModel>(throwOnFailure: true, cancellationToken).ConfigureAwait(true);
```

Comment:
You don't need to retrieve this here, your package unneccesarily waits on the MEF cache to read or be rebuilt here, just retrieve at exact momement you need.

Reply:
I started going down that path, and maybe it's something to pursue, but there are a lot of consumers of the ComponentModel property, and making them async wasn't going to be simple

Reply:
Took another look, and this is not going to be a simple thing to defer (for long, at least). 

For example, in RoslynPackage.InitializeAsync, it calls base.InitializeAsync (which sets _componentModel_doNotAccessDirectly), and then two of the next three statements end up accessing ComponentModel and calling methods on the items returned from the MEF catalog.

Reply:
Took some stopwatch data. About 500 ms of the 700 ms I see in InitializeAsync is due to obtaining MEF. Need to think on this more.

Reply:
That's when its up-to-date that's 15 seconds+ on a full rebuild.

Reply:
Good to know.  Let's bite off pieces as we can. I'll be coming back from vacation on Monday and can help Todd out with this. 

We'll likely have several PRs to get through the bulk of the work, depending on how invasive the changes are. 

@toddgrunke feel free to open an issue to help track upcoming immediate work. Thanks!

Reply:
Also, we should def be measuring *both* "warm Mef" and "full Mef rebuild time"


@davekean how possible would it be to have dedicated speedometer tests for both?  It would be great for measuring our progress here over time, and especially good for preventing regressions. 

If we already have tests here (say... Time to open a basic project, and time to open a single file in a project), could you link is to those? 

Thanks!

Reply:
@davkean -- would moving the usage of MEF out of InitializeAsync and instead use it in OnAfterPackageLoadedAsync and IVsEditorFactory.CreateEditorInstance be a sufficient deferral of MEF usage?